docs/design/features/byreflike-generics.md

```diff
@@ -0,0 +1,75 @@
+# Generics parameters of ByRefLike types
+
+Using ByRefLike types in Generic parameters is possible by building upon support added for `ref` fields. Scenarios that would benefit most from this are those involving `Span<T>`. For example, consider the following examples:
+
+- `Span<TypedReference>` &ndash; Represents the general case where a ByRefLike type is used as a Generic parameter. This specific case would be desirable for a more efficient Reflection API.
+- `Span<Span<char>>` &ndash; Nested `Span<T>` types would be of benefit in the parsing result of strings.
+
+## Runtime impact
+
+Supporting ByRefLike type as Generic parameters will impact the following IL instructions:
+
+- `box` &ndash; Types with ByRefLike parameters used in fields cannot be boxed.
+- `throw` &ndash; Requires an object reference on the stack so not directly impacted since boxing is not permitted.
+- `stsfld` / `ldsfld` &ndash; Type fields of a ByRefLike parameter cannot be marked `static`.
+- `newarr` / `stelem` / `ldelem` / `ldelema` &ndash; Arrays are not able to contain ByRefLike types.
+    - `newobj` &ndash; For multi-dimensional array construction.
+- `constrained.callvirt` &ndash; This IL sequence must resolve to a method implemented on `object`, or a default interface method.
+- Use of any Generic which doesn’t expect a ByRefLike parameter as a Generic parameter.
+
+## Proposal
+
+A new Attribute API will be defined to indicate which Generic parameters are permissible to be of any type&mdash;including ByRefLike types. The Attribute could be used by the compiler to implement a generic non-constraint.
+
+```csharp
+namespace System.Runtime.CompilerServices
+{
+    [AttributeUsage(AttributeTargets.Class | AttributeTargets.Method, AllowMultiple = true)]
```

Comment:
`| AttributeTargets.Struct | AttributeTargets.Delegate | AttributeTargets.Interface`

Reply:
According to the proposal in dotnet/csharplang#5817 generic delegates would automatically support byref-like types. Why would applying the attribute on delegates still be needed?

Reply:
> generic delegates would automatically support byref-like types

I don't see that in the proposal nor would that be possible at present. There is no way to support ByRefLike types in Generics in any form at present. Perhaps I am misunderstanding your point though.

Reply:
Oops, I misclicked. The proposal is dotnet/csharplang#5689 and the relevant point is https://github.com/dotnet/csharplang/pull/5689/files#diff-67b62b3a3e0ed8896aa17e6a455ddc5ca90586364864c7bfd280223999a98b3fR191.

Reply:
> The proposal is https://github.com/dotnet/csharplang/pull/5689

Thanks.

> Why would applying the attribute on delegates still be needed?

Because the compiler still needs a way to convey that message and encode it. At present there is no way to encode that case and this attribute is an option. However, others exist - see various statements in this design doc by @jkotas and @MichalStrehovsky.

Reply:
Moving indication to metadata flags.